### PR TITLE
HarfBuzz

### PR DESCRIPTION
**Pull request by zauguin**
Continuation of #97.

What is still needed to merge this? (Feel free to extend this list)

 - [X] **Tests** Can we do LuaHBTeX tests on Travis? Things which should be covered by tests:
    - [ ] *Complicated shaping* e.g. a Devanagari test? That should cover most edge-cases.
    - [x] *Font feature handling* especially `tlig`
    - [x] *Discretionary nodes* Especially interactions with Ligatures, e.g. the `of\-f\-ice` example from the manual.
    - [x] *Node leaks* I think we should add, at least for some tests, messages how many nodes are allocated to avoid leaking them by accident.
    - [X] *Color fonts* At least ~one `glyf`~, one `COLR` and one `CBDT` font.
    - [x] *Interaction with other fonts* Do HarfBuzz fonts properly ignore non-HarfBuzz fonts.
 -  ~**Documentation** I think we mostly need a note that `mode=harf` can be specified, maybe some HarfBuzz specific options. Currently `shapers` and `palette` are the only ones I can think of.~ Moved to #111
 - ~Related to tests: Theoretically this should also do e.g. Graphite2 shaping, but I never tried it and don't even know which fonts to use there. I'll look into it but I would expect that it just works (TM)~ Doesn't "just work", moved to #112.
 - [x] We are currently triggering https://github.com/ufyTeX/luaharfbuzz/pull/44 which is fixed upstream but the fix isn't released yet, so maybe we should add a work-around.

### CODE REVIEW

**Comment by khaledhosny** [comment]
Re-tests, I think it would be nice to check the test files I had to make sure nothing broke. They look random and unpolished, but I tried my best to have as wider coverage as possible especially for trickier parts of the code.

**Comment by khaledhosny** [comment]
I also had tests for graphite fonts, testing AAT fonts would be nice as well (there are not that many free fonts that have AAT tables, so the tests would run only on macOS and use system AAT fonts).

File: src/luaotfload-auxiliary.lua [hunk 1]
```diff
@@ -477,8 +512,36 @@ function aux.provides_language(font_id, asked_script, asked_language)
...
+    return false
+  end
+  local hbface = get_hbface(tfmdata)
+  if hbface then
+    asked_language  = stringupper(asked_language)
+    if asked_language == "DFLT" then
+      return aux.provides_script(font_id, asked_script)
+    end
+    local script = harf.Tag.new(asked_script)
+    -- HarfBuzz expects "DFLT" in lower-case
```
Comment (khaledhosny) [inline src/luaotfload-auxiliary.lua]: Not HarfBuzz but the OpenType spec, there is no `dflt` script tag.

Comment (khaledhosny) [reply src/luaotfload-auxiliary.lua]: https://docs.microsoft.com/en-us/typography/opentype/spec/scripttags

Comment (zauguin) [reply src/luaotfload-auxiliary.lua]: @khaledhosny  Right, the comment moved to the wrong location, it was supposed to be above `asked_language` where the specs do not name the default.

File: src/luaotfload-auxiliary.lua [hunk 2]
```diff
@@ -558,12 +620,31 @@ function aux.provides_feature(font_id,        asked_script,
...
+  if not tfmdata then
+    logreport ("log", 0, "aux", "no font with id %d", font_id)
+    return false
+  end
+  local hbface = get_hbface(tfmdata)
+  if hbface then
+    asked_language  = stringupper(asked_language)
+    local script = harf.Tag.new(asked_script)
+    -- HarfBuzz expects "DFLT" in lower-case
+    local language = harf.Tag.new(asked_language == "DFLT" and "dflt" or asked_language)
```
Comment (khaledhosny) [inline src/luaotfload-auxiliary.lua]: It is the otherway around here, the language tag should be `dflt`, though in this case it is a convention since the default language does not have an actual tag and is stored in the font binary in a special record.

File: src/luaotfload-harf-define.lua [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+  [0x2013] = { [0x002D] = { char = 0x2014 } }, -- [---]
+  [0x002D] = { [0x002D] = { char = 0x2013 } }, -- [--]
+  [0x0060] = { [0x0060] = { char = 0x201C } }, -- [``]
+  [0x0027] = { [0x0027] = { char = 0x201D } }, -- ['']
+  [0x0021] = { [0x0060] = { char = 0x00A1 } }, -- [!`]
+  [0x003F] = { [0x0060] = { char = 0x00BF } }, -- [?`]
+  [0x002C] = { [0x002C] = { char = 0x201E } }, -- [,,]
+  [0x003C] = { [0x003C] = { char = 0x00AB } }, -- [<<]
+  [0x003E] = { [0x003E] = { char = 0x00BB } }, -- [>>]
+}
```
Comment (khaledhosny) [inline src/luaotfload-harf-define.lua]: I used global variables for `tlig` and `trep` with the intention that people would use them to implement ligatures in a similar (but limited) way to what the ConTeXt based loader provides. I’m not too attached to this, but I recall many queries from people wanting to patch fonts on the fly and add missing ligature records. I think it would be nice to keep this functionality in some way or another.

File: src/luaotfload-harf-define.lua [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+  if rawfeatures.language then
+    spec.language = hb.Language.new(rawfeatures.language)
+  end
+  if rawfeatures.script then
+    spec.script = hb.Script.new(rawfeatures.script)
+  end
+  for key, val in next, rawfeatures do
+    if key:len() == 4 then
+      -- 4-letter options are likely font features, but not always, so we do
+      -- some checks below. We put non feature options in the `options` dict.
```
Comment (khaledhosny) [inline src/luaotfload-harf-define.lua]: This comments does not reflect what the code is currently doing.

File: src/luaotfload-harf-define.lua [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+  [0x003C] = { [0x003C] = { char = 0x00AB } }, -- [<<]
+  [0x003E] = { [0x003E] = { char = 0x00BB } }, -- [>>]
+}
+
+local function scalefont(data, spec)
+  local size = spec.size
+  local features = spec.features.normal
+  features.mode = 'plug'
+  features.features = 'harf'
+  fonts.constructors.checkedfeatures("otf", features)
```
Comment (khaledhosny) [inline src/luaotfload-harf-define.lua]: Is this the code that adds script-specific features? If so it shouldn’t be done here as HarfBuzz maintains this internally and manually enabling features that require text analysis by the engine (e.g. Arabic `init`) will activate it globally for the buffer and give the wrong result.

File: src/luaotfload-harf-plug.lua [hunk 6]
```diff
@@ -0,0 +1,1000 @@
...
+      font.addcharacters(fontid, { characters = newcharacters })
+    end
+  end
+end
+
+-- FIXME: Move this into generic parts of luaotfload
+local function get_glyph_info(n)
+  local props = properties[todirect(n)]
+  if not props then return end
+  return props and props.glyph_info or nil
```
Comment (khaledhosny) [inline src/luaotfload-harf-plug.lua]: Since LuaTeX slightly changed the semantics of the callback, and now `nil` will cause, it to print the node `character`, this code needs to be adapted and return `nil` only when it is not a HarfBuzz processed font, otherwise return empty sting (this is to handle characters that map to multiple glyphs, the first glyph will print the character and the rest should print nothing).

Comment (zauguin) [reply src/luaotfload-harf-plug.lua]: That's handled earlier: The remaining characters have `props.glyph_info` set to the empty string, so for them nothing is printed. The function still has to change to print something sensible for other fonts though.

File: src/luaotfload-harf-define.lua (hunk 5 above)
Comment (zauguin) [reply src/luaotfload-harf-define.lua]: This adds default features, not script specific default features. This is necessary to properly interact with some user-defined features. Anyway, this does not interact with HarfBuzz features at all: They are extracted earlier and only look at `features.raw`. That table only contains the features passed by the user, not automatic features.

File: src/luaotfload-harf-plug.lua [hunk 7]
```diff
@@ -0,0 +1,1000 @@
...
+  for k, v in next, base_callback_descriptions(name) do
+    saved_callbacks[k] = { base_remove_from_callback(name, v) }
+  end
+  base_add_to_callback(name, func, "Harf "..name.." callback")
+  for _, v in next, saved_callbacks do
+    base_add_to_callback(name, v[1], v[2])
+  end
+end
+
+add_to_callback('pre_output_filter', post_process_nodes) -- FIXME: Wrong callback, but I want to get rid of the whole function anyway
```
Comment (khaledhosny) [inline src/luaotfload-harf-plug.lua:976]: What is wrong with this callback? The code is placed here to be run after line breaking, otherwise you end with unbalanced actual text tags because some nodes went away during line breaking.

File: src/luaotfload-harf-define.lua (hunk 3 above)
Comment (zauguin) [reply src/luaotfload-harf-define.lua]: This can be implemented using the fontloaders mechanism for user defined features. These tables became local especially to avoid modifications: User-defined ligatures should in general not be controlled by `tlig` but by a custom feature, otherwise they will often affect more fonts than intended.

File: src/luaotfload-harf-plug.lua (hunk 7 above)
Comment (zauguin) [reply src/luaotfload-harf-plug.lua]: The alternatives would be `post_linebreak_filter` and `hpack_filter` similar to the color handling or a potential future pre-shipout hook.

File: src/luaotfload-harf-plug.lua [hunk 8]
```diff
@@ -0,0 +1,1000 @@
...
+local function post_process_nodes(head, groupcode)
+  return tonode(post_process(todirect(head)))
+end
+
+local function run_cleanup()
+  -- Remove temporary PNG files that we created, if any.
+  -- FIXME: It would be nice if we wouldn't need this
+  for _, path in next, pngcachefiles do
+    osremove(path)
+  end
```
Comment (khaledhosny) [inline src/luaotfload-harf-plug.lua]: luaotfload could cache the PNGs inside its font cache directories, but it would then need to handle the case when the font changes.

File: src/luaotfload-harf-plug.lua [hunk 9]
```diff
@@ -0,0 +1,1000 @@
...
+    -- local file = open(path, "wb"):write():close()
+    -- file:write(data)
+    -- file:close()
+  i = img.scan{filename = path}
+  pngcache[hash] = i
+end
+return i
+end
+
+local function get_png_glyph(gid, fontid, characters, haspng)
```
Comment (khaledhosny) [inline src/luaotfload-harf-plug.lua]: Somehow the code from here on lost one level of indentation.

File: src/luaotfload-harf-define.lua (hunk 3 above)
Comment (khaledhosny) [reply src/luaotfload-harf-define.lua]: I agree, but there is no other mechanism for HarfBuzz-processed fonts, any code that modifies the Lua tables that are used by the other loader will not work here. Another table/option/feature is needed, it does not have to be `tlig`.

Comment (zauguin) [reply src/luaotfload-harf-define.lua]: The fontloader provides multiple kinds of user-defined features, there are the higher-level ones which reuse the fontloaders OpenType feature handling and there is are features on a lower level which can modify the font table or modify each nodelist which contains the font. The last mechanism is e.g. used in luaotfload-colors.lua to implement the `color` feature. These low-level feature work with HarfBuzz fonts too.

These could be used to implement something like `tlig` by modifying the font table to add ligatures exactly like it is done here and `trep` could be implemented by pre-processing the node list and rewriting the affected nodes.

We could add a nicer interface for this.

File: src/luaotfload-harf-plug.lua (hunk 9 above)
Comment (zauguin) [reply src/luaotfload-harf-plug.lua]: Fixed

File: src/luaotfload-harf-plug.lua [hunk 10]
```diff
@@ -993,10 +996,11 @@ local function set_tounicode()
...
 
 -- FIXME: Move this into generic parts of luaotfload
+local utfchar = utf8.char
 local function get_glyph_info(n)
-  local props = properties[todirect(n)]
-  if not props then return end
-  return props and props.glyph_info or nil
+  n = todirect(n)
+  local props = properties[n]
+  return props and props.glyph_info or utfchar(getchar(n))
```
Comment (khaledhosny) [inline src/luaotfload-harf-plug.lua]: I think returning `nil` here was fine, it will cause the engine to do the equivalent of `utfchar(getchar(n))` which might be slightly faster.

File: src/luaotfload-harf-define.lua [hunk 11]
```diff
@@ -350,10 +353,14 @@ fonts.readers.harf = function(spec)
   spec.hb_features = hb_features
 
   if rawfeatures.language then
-    spec.language = hb.Language.new(rawfeatures.language)
+    local language = stringupper(rawfeatures.language)
+    spec.language = hb.Language.new(language == "DFLT" and "dflt"
+                                                  or script)
```
Comment (khaledhosny) [inline src/luaotfload-harf-define.lua]: I think this should be `or language`.

File: src/luaotfload-harf-plug.lua (hunk 10 above)
Comment (zauguin) [reply src/luaotfload-harf-plug.lua]: No, if the callback exists but returns nil the engine uses `"` followed by the hexadecimal representation.

File: src/luaotfload-harf-define.lua (hunk 11 above)
Comment (zauguin) [reply src/luaotfload-harf-define.lua]: 🙈 thanks!

File: src/luaotfload-harf-plug.lua (hunk 10 above)
Comment (khaledhosny) [reply src/luaotfload-harf-plug.lua]: I see.

**Comment by zauguin** [comment]
> Re-tests, I think it would be nice to check the test files I had to make sure nothing broke. They look random and unpolished, but I tried my best to have as wider coverage as possible especially for trickier parts of the code.

Thank you, the tests there are great. I fixed all bugs I found using them and we should try to convert them into our regular test format.

> I also had tests for graphite fonts, testing AAT fonts would be nice as well (there are not that many free fonts that have AAT tables, so the tests would run only on macOS and use system AAT fonts).

That's a bit more complicated because AFAICT neither Ulrike, Travis nor me have macOS systems available, so while we can one-time tests of this functionality, regular, automated tests don't really work. But I think if graphite and HarfBuzz's normal ot shaping works, other HarfBuzz shapers shouldn't make a problem because the used shaper shouldn't change HarfBuzz's interface.

**Comment by josephwright** [comment]
https://docs.travis-ci.com/user/reference/osx/

**Comment by zauguin** [comment]
> https://docs.travis-ci.com/user/reference/osx/

I stand corrected.

**Comment by khaledhosny** [comment]
HarfBuzz AAT support is cross-platform as well, the problem is finding freely licensed AAT fonts, so the simplest solution is to run on Travis macOS.

**Comment by zauguin** [comment]
> HarfBuzz AAT support is cross-platform as well, the problem is finding freely licensed AAT fonts, so the simplest solution is to run on Travis macOS.

Thanks, I thought that AAT is disabled on non-macOS systems because HarfBuzz didn't have a AAt renderer on any system I tested, but now I realized that AAt is part of the ot renderer.

**Comment by khaledhosny** [comment]
For the tests, I was under the impression that some of the things I test for can only be captured in the PDF output. Unfortunately I don’t remember the details, and I might have been mistaken.

**Comment by zauguin** [comment]
@khaledhosny I am not sure if these are the things you thought about, but I am currently aware of basically five places where pure Log file based testing has problems here:

  * Given that we overwrite which the glyph information that is written to the log file through `get_glyph_info`, we need a test which ensures that we actually write the glyphs we want to write. (This doesn't require PDF tests, but we have to modify the `get_glyph_info` callback for some tests.)
  * We need to test that the font information is correct, so we need to print the font table to the log or do PDF testing. Given that the font tables are modified using `addcharacters`, some PDF tests are required.
  * Related to that, one special part of font tables the new handling for color fonts uses are virtual fonts. They are not visible from TeX and are applied after `\tracingoutput`, so they can only be detected through PDF tests.
  * The current code inserts ActualText in a way which doesn't print the text to the log but only `<lua data referece xxx>` for some integer xxx, so we need at leat one test which ensures that these `\pdfliteral`s contain te right values. This isn't a big problem though, because of the first point: The value written there  pretty much corresponds to the get_glyph_value of the first glyph and that *is* checked. So we only need one test to ensure that this equivalence holds and need to keep the `get_glyph_info` for most tests.
  * Additionally some glyph data isn't printed in the log, especially `xoffset` and `yoffset`. (Most other values indirectly appear in the log)

So where does this leave us? We can use an appropriate `get_glyph_info` callback which both includes the output of the original `get_glyph_info` to test that and includes all additional info we would like to test, effectively solving the first, fourth(partially) and fifth point. Additionally we need PDF tests at some positions:

  * The color font tests, both `colored.lvt` and `pngemoji.lvt` need PDF testing.
  * We need an additional test which loads some fonts, uses some clusters and uses PDF and Log file testing to ensure that the second point is handled and the fourth doesn't become a problem.

Did I miss something?

**Comment by khaledhosny** [comment]
Handling of font metrics (ascender, descender, x-height, cap-height etc) used to  be an issue as well since LuaTeX don’t read these from the relevant OpenType tables and instead “guess” them from the metrics of certain characters, and if the glyphs are not loaded early on this would break. This often affects line height, but I’m not sure testing line height is enough and whether other changes would affect PDF rendering.

There is also the `/ToUnicode` stuff in the font CMAP that is only visible in the PDF file (this probably affects many of the tests, since it is a rather hairy code that breaks with the even innocent looking changes)..

**Comment by zauguin** [comment]
> Handling of font metrics (ascender, descender, x-height, cap-height etc) used to be an issue as well since LuaTeX don’t read these from the relevant OpenType tables and instead “guess” them from the metrics of certain characters, and if the glyphs are not loaded early on this would break. This often affects line height, but I’m not sure testing line height is enough and whether other changes would affect PDF rendering.

Right, but given that we load a full font that's not a problem and is checked as a side-effect of any PDF test. Also the most interesting case is what happens if the affected characters are not used in the document, so this is rather well covered in the current version.

> 
> There is also the `/ToUnicode` stuff in the font CMAP that is only visible in the PDF file (this probably affects many of the tests, since it is a rather hairy code that breaks with the even innocent looking changes)..

Actually I think that code has two parts: The part where the string representation is determined and the part where it is used to fill tounicode or ActualText.
The first part is quite fragile, but it is well covered by the log tests while the second part is actually quite straight forward code which only has a few cases to test, so this doesn't need to be checked with every test.

**Comment by khaledhosny** [comment]
@zauguin OK.

**Comment by zauguin** [comment]
@u-fischer With the current version, take any file with color fonts (especially the layered ones in `colored.lvt`) or try `arabic.lvt`.
